core/src/services/libsql/backend.rs

```diff
@@ -0,0 +1,345 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::collections::HashMap;
+use std::fmt::Debug;
+
+use async_trait::async_trait;
+use libsql_client::{args, Config, Statement, SyncClient, Value};
```

Comment:
`SyncClient` is calling `futures::executor::block_on`, we can't use them in OpenDAL. Instead, we can use `Client` directly and let users to use `blocking_layer`.

Reply:
`Client` does not implement the `Sync` trait, so it will report an error when using it, how should I solve it?

<details> <summary>msg</summary>

```plaintext
error: future cannot be sent between threads safely
   --> core/src/services/libsql/backend.rs:286:65
    |
286 |       async fn set(&self, path: &str, value: &[u8]) -> Result<()> {
    |  _________________________________________________________________^
287 | |
288 | |         let client = Client::from_config(self.get_config()?)
289 | |             .await
...   |
302 | |         Ok(())
303 | |     }
    | |_____^ future created by async block is not `Send`
    |
    = help: within `libsql_client::Client`, the trait `Sync` is not implemented for `*mut libsql_sys::ffi::sqlite3`
note: future is not `Send` as this value is used across an await
   --> core/src/services/libsql/backend.rs:299:13
    |
297 |         client
    |         ------ has type `&libsql_client::Client` which is not `Send`
298 |             .execute(Statement::with_args(query, args!(path, value.to_vec())))
299 |             .await
    |             ^^^^^^ await occurs here, with `client` maybe used later
300 |             .map_err(|err| Error::new(ErrorKind::Unexpected, "set failed").set_source(err))?;
    |                                                                                             - `client` is later dropped here
help: consider moving this into a `let` binding to create a shorter lived borrow
   --> core/src/services/libsql/backend.rs:297:9
    |
297 | /         client
298 | |             .execute(Statement::with_args(query, args!(path, value.to_vec())))
    | |______________________________________________________________________________^
    = note: required for the cast from `[async block@core/src/services/libsql/backend.rs:286:65: 303:6]` to the object type `dyn futures::Future<Output = std::result::Result<(), types::error::Error>> + std::marker::Send`

```
</details>

Reply:
You can wrap a `Arc<Mutex<T>>` for it.

Reply:
BTW, libsql_client looks like a simple wrapper around reqwest. Maybe we can build request by ourselves?

https://docs.rs/libsql-client/latest/src/libsql_client/reqwest.rs.html#17-39

Reply:
> BTW, libsql_client looks like a simple wrapper around reqwest. Maybe we can build request by ourselves?
> 
> https://docs.rs/libsql-client/latest/src/libsql_client/reqwest.rs.html#17-39

I didn't find standard api documentation for the libsql service, and I realized that the libsql sdk from a few months ago doesn't work with the latest version of libsql anymore. Would it take more effort to maintain api changes if we build http requests directly?

Reply:
Only the `/queries` api is mentioned here, but the `/v2/pipeline` api is used in libsql_client.

https://github.com/libsql/sqld/blob/main/docs/http_api.md

Reply:
It seems that libsql only provides one API, and we can use `hrana_client_proto` for its `hrana_client_proto::pipeline::ServerMsg`. I'm fine with either of them. Perhaps starting with `Arc<Mutex<T>>` is sufficient.

Reply:
> You can wrap a `Arc<Mutex<T>>` for it.

Hi @Xuanwo, I tried it and still got the same error after using the Arc<Mutex< Client >>.

<details> <summary>msg</summary>

```plaintext
error: future cannot be sent between threads safely
   --> core/src/services/libsql/backend.rs:310:65
    |
310 |       async fn set(&self, path: &str, value: &[u8]) -> Result<()> {
    |  _________________________________________________________________^
311 | |         let query = format!(
312 | |             "INSERT OR REPLACE INTO `{}` (`{}`, `{}`) VALUES (?, ?)",
313 | |             self.table, self.key_field, self.value_field
...   |
323 | |         Ok(())
324 | |     }
    | |_____^ future created by async block is not `Send`
    |
    = help: within `libsql_client::Client`, the trait `Sync` is not implemented for `*mut libsql_sys::ffi::sqlite3`
note: future is not `Send` as this value is used across an await
   --> core/src/services/libsql/backend.rs:321:13
    |
319 |         client.lock().await
    |         ------------------- has type `&libsql_client::Client` which is not `Send`
320 |             .execute(Statement::with_args(query, args!(path, value.to_vec())))
321 |             .await
    |             ^^^^^^ await occurs here, with `client.lock().await` maybe used later
322 |             .map_err(|err| parse_error_msg(err, "set failed"))?;
    |                                                                - `client.lock().await` is later dropped here
help: consider moving this into a `let` binding to create a shorter lived borrow
   --> core/src/services/libsql/backend.rs:319:9
    |
319 | /         client.lock().await
320 | |             .execute(Statement::with_args(query, args!(path, value.to_vec())))
    | |______________________________________________________________________________^
    = note: required for the cast from `[async block@core/src/services/libsql/backend.rs:310:65: 324:6]` to the object type `dyn futures::Future<Output = std::result::Result<(), types::error::Error>> + std::marker::Send`


```

</details>

Reply:
You will need to write code like the following:

```rust
// This block is required to make sure guard is dropped.
{
   let guard = client.lock().await;
   guard
     .execute(Statement::with_args(query, args!(path, value.to_vec())))
     .await
     .map_err(|err| parse_error_msg(err, "set failed"))?;
}

// guard has been dropeed, so we make sure that we don't hold this guard across await point
```

Reply:
It's still not working.🥺

![err](https://github.com/apache/incubator-opendal/assets/38717659/6e28c037-4212-4406-aaad-3203a4007c0b)